File: src/ccc/store/factory.py
"""Build a tiered ContextStore: cache (Redis -> SQLite) + optional backing."""

from __future__ import annotations

from dataclasses import replace
from typing import cast

from ccc.store.adapters.mongodb import open_mongo
from ccc.store.adapters.mysql import open_mysql
from ccc.store.adapters.postgres import open_postgres
from ccc.store.adapters.redis import open_redis
from ccc.store.adapters.sqlite import open_sqlite
from ccc.store.config import StoreConfig, load_config
from ccc.store.types import BackendKind, ContextStore, ProbeAttempt, StoreInfo


def _try_open_cache(
    cfg: StoreConfig, attempts: list[ProbeAttempt]
) -> ContextStore:
    if cfg.redis_url and not cfg.disable_redis:
        try:
            store = open_redis(
                url=cfg.redis_url,
                namespace=cfg.namespace,
                probe_timeout_ms=cfg.probe_timeout_ms,
            )
            attempts.append(ProbeAttempt(kind="redis", ok=True))
            return store
        except Exception as exc:
            attempts.append(
                ProbeAttempt(kind="redis", ok=False, error=str(exc))
            )
    if not cfg.disable_sqlite_memory:
        try:
            store = open_sqlite(
                path=":memory:",
                namespace=cfg.namespace,
                kind="sqlite-memory",
            )
            attempts.append(ProbeAttempt(kind="sqlite-memory", ok=True))
            return store
        except Exception as exc:
            attempts.append(
                ProbeAttempt(kind="sqlite-memory", ok=False, error=str(exc))
            )
    store = open_sqlite(
        path=cfg.sqlite_path,
        namespace=cfg.namespace,
        kind="sqlite-disk",
    )
    attempts.append(ProbeAttempt(kind="sqlite-disk", ok=True))
    return store


def _try_open_backing(
    cfg: StoreConfig, attempts: list[ProbeAttempt]
) -> ContextStore | None:
    if cfg.mysql_url:
        store = open_mysql(
            url=cfg.mysql_url,
            namespace=cfg.namespace,
            table=cfg.table_name,
            probe_timeout_ms=cfg.probe_timeout_ms,
        )
        attempts.append(ProbeAttempt(kind="mysql", ok=True))
        return store
    if cfg.postgres_url:
        store = open_postgres(
            url=cfg.postgres_url,
            namespace=cfg.namespace,
            table=cfg.table_name,
            probe_timeout_ms=cfg.probe_timeout_ms,
        )
        attempts.append(ProbeAttempt(kind="postgres", ok=True))
        return store
    if cfg.mongo_url:
        store = open_mongo(
            url=cfg.mongo_url,
            namespace=cfg.namespace,
            collection=cfg.table_name,
            probe_timeout_ms=cfg.probe_timeout_ms,
        )
        attempts.append(ProbeAttempt(kind="mongodb", ok=True))
        return store
    return None


class TieredStore(ContextStore):
    """Cache-fronted store: reads cache first, fills from backing on miss; writes to both."""

    def __init__(self, cache: ContextStore, backing: ContextStore | None) -> None:
        self._cache = cache
        self._backing = backing
        self.kind = cache.kind

    def get(self, key: str) -> str | None:
        cached = self._cache.get(key)
        if cached is not None:
            return cached
        if self._backing is None:
            return None
        persisted = self._backing.get(key)
        if persisted is not None:
            self._cache.set(key, persisted)
        return persisted

    def set(self, key: str, value: str, *, ttl_ms: int | None = None) -> None:
        self._cache.set(key, value, ttl_ms=ttl_ms)
        if self._backing is not None:
            self._backing.set(key, value, ttl_ms=ttl_ms)

    def delete(self, key: str) -> bool:
        a = self._cache.delete(key)
        b = self._backing.delete(key) if self._backing is not None else False
        return a or b

    def list(self, prefix: str | None = None) -> list[str]:
        seen = set(self._cache.list(prefix))
        if self._backing is not None:
            seen.update(self._backing.list(prefix))
        return sorted(seen)

    def close(self) -> None:
        self._cache.close()
        if self._backing is not None:
            self._backing.close()


def create_store(
    config: StoreConfig | None = None,
) -> tuple[ContextStore, StoreInfo]:
    """Open a tiered store. Returns (store, info) where info reports probes."""
    cfg = load_config(config)
    attempts: list[ProbeAttempt] = []
    cache = _try_open_cache(cfg, attempts)
    backing = _try_open_backing(cfg, attempts)
    store: ContextStore = TieredStore(cache, backing) if backing else cache
    info = StoreInfo(
        cache=cast(BackendKind, cache.kind),
        backing=cast("BackendKind | None", backing.kind if backing else None),
        attempts=tuple(attempts),
    )
    # Quiet the unused-import warning on `replace` while leaving the import
    # available for callers who manipulate StoreConfig copies.
    _ = replace
    return store, info

File: context-store/src/factory.ts
import type { BackendKind, ContextStore, SetOptions, StoreConfig } from "./types.js";
import { loadConfig } from "./config.js";
import { openSqlite } from "./adapters/sqlite.js";
import { openRedis } from "./adapters/redis.js";
import { openMysql } from "./adapters/mysql.js";
import { openPostgres } from "./adapters/postgres.js";
import { openMongo } from "./adapters/mongodb.js";

export interface StoreInfo {
  cache: BackendKind;
  backing: BackendKind | null;
  attempts: Array<{ kind: BackendKind; ok: boolean; error?: string }>;
}

export interface ResolvedStore {
  store: ContextStore;
  info: StoreInfo;
}

async function tryOpenCache(
  cfg: ReturnType<typeof loadConfig>,
  attempts: StoreInfo["attempts"],
): Promise<ContextStore> {
  if (cfg.redisUrl && !cfg.disableRedis) {
    try {
      const s = await openRedis({
        url: cfg.redisUrl,
        namespace: cfg.namespace,
        probeTimeoutMs: cfg.probeTimeoutMs,
      });
      attempts.push({ kind: "redis", ok: true });
      return s;
    } catch (e) {
      attempts.push({
        kind: "redis",
        ok: false,
        error: (e as Error).message,
      });
    }
  }
  if (!cfg.disableSqliteMemory) {
    try {
      const s = openSqlite({
        path: ":memory:",
        namespace: cfg.namespace,
        kind: "sqlite-memory",
      });
      attempts.push({ kind: "sqlite-memory", ok: true });
      return s;
    } catch (e) {
      attempts.push({
        kind: "sqlite-memory",
        ok: false,
        error: (e as Error).message,
      });
    }
  }
  const s = openSqlite({
    path: cfg.sqlitePath,
    namespace: cfg.namespace,
    kind: "sqlite-disk",
  });
  attempts.push({ kind: "sqlite-disk", ok: true });
  return s;
}

async function tryOpenBacking(
  cfg: ReturnType<typeof loadConfig>,
  attempts: StoreInfo["attempts"],
): Promise<ContextStore | null> {
  if (cfg.mysqlUrl) {
    const s = await openMysql({
      url: cfg.mysqlUrl,
      namespace: cfg.namespace,
      table: cfg.tableName,
      probeTimeoutMs: cfg.probeTimeoutMs,
    });
    attempts.push({ kind: "mysql", ok: true });
    return s;
  }
  if (cfg.postgresUrl) {
    const s = await openPostgres({
      url: cfg.postgresUrl,
      namespace: cfg.namespace,
      table: cfg.tableName,
      probeTimeoutMs: cfg.probeTimeoutMs,
    });
    attempts.push({ kind: "postgres", ok: true });
    return s;
  }
  if (cfg.mongoUrl) {
    const s = await openMongo({
      url: cfg.mongoUrl,
      namespace: cfg.namespace,
      collection: cfg.tableName,
      probeTimeoutMs: cfg.probeTimeoutMs,
    });
    attempts.push({ kind: "mongodb", ok: true });
    return s;
  }
  return null;
}

class TieredStore implements ContextStore {
  readonly kind: BackendKind;
  constructor(
    private cache: ContextStore,
    private backing: ContextStore | null,
  ) {
    this.kind = cache.kind;
  }
  async get(key: string): Promise<string | null> {
    const cached = await this.cache.get(key);
    if (cached !== null) return cached;
    if (!this.backing) return null;
    const persisted = await this.backing.get(key);
    if (persisted !== null) {
      await this.cache.set(key, persisted);
    }
    return persisted;
  }
  async set(key: string, value: string, opts?: SetOptions): Promise<void> {
    await this.cache.set(key, value, opts);
    if (this.backing) await this.backing.set(key, value, opts);
  }
  async delete(key: string): Promise<boolean> {
    const a = await this.cache.delete(key);
    const b = this.backing ? await this.backing.delete(key) : false;
    return a || b;
  }
  async list(prefix?: string): Promise<string[]> {
    const cacheKeys = await this.cache.list(prefix);
    if (!this.backing) return cacheKeys;
    const backingKeys = await this.backing.list(prefix);
    return Array.from(new Set([...cacheKeys, ...backingKeys])).sort();
  }
  async close(): Promise<void> {
    await this.cache.close();
    if (this.backing) await this.backing.close();
  }
}

export async function createStore(
  config: StoreConfig = {},
): Promise<ResolvedStore> {
  const cfg = loadConfig(config);
  const attempts: StoreInfo["attempts"] = [];
  const cache = await tryOpenCache(cfg, attempts);
  const backing = await tryOpenBacking(cfg, attempts);
  const store = backing ? new TieredStore(cache, backing) : cache;
  return {
    store,
    info: {
      cache: cache.kind,
      backing: backing?.kind ?? null,
      attempts,
    },
  };
}
